packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesCodeSigningConfiguration.swift

```diff
@@ -0,0 +1,138 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+import Foundation
+import CommonCrypto
+import ASN1Decoder
+
+@objc public enum EXUpdatesCodeSigningConfigurationError : Int, Error {
+  case CertificateParseError
+  case CertificateValidityError
+  case CertificateDigitalSignatureNotPresentError
+  case CertificateMissingCodeSigningError
+  case KeyIdMismatchError
+  case PublicKeyInvalidError
+  case SignatureEncodingError
+}
+
+struct EXUpdatesCodeSigningMetadataFields {
+  static let KeyIdFieldKey = "keyid"
+  static let AlgorithmFieldKey = "alg"
+}
+
+@objc
+public class EXUpdatesCodeSigningConfiguration : NSObject {
+  static let EXUpdatesCryptoRootPublicKeyTag = "exp.host.rootkey"
+  
+  // ASN.1 path to the extended key usage info within a CERT
+  static let EXUpdatesCodeSigningCertificateExtendedUsageCodeSigningOID = "1.3.6.1.5.5.7.3.3"
+  
+  private var certificateDataDer: Data
+  private var keyId: String
+  private var algorithm: EXUpdatesCodeSigningAlgorithm
+  
+  @objc
+  public required init(certificate: String, metadata: [String: String]) throws {
+    guard let certificateData = certificate.data(using: .utf8) else { throw EXUpdatesCodeSigningConfigurationError.CertificateParseError }
+    
+    guard let certificateDataDer = EXUpdatesCodeSigningConfiguration.decodeToDER(pem: certificateData) else {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateParseError
+    }
+    self.certificateDataDer = certificateDataDer
+    
+    let x509Certificate = try X509Certificate(der: certificateDataDer)
+    
+    guard x509Certificate.checkValidity() else {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateValidityError
+    }
+    
+    let keyUsage = x509Certificate.keyUsage
+    if (x509Certificate.keyUsage.isEmpty || !keyUsage[0]) {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateDigitalSignatureNotPresentError
+    }
+    
+    let extendedKeyUsage = x509Certificate.extendedKeyUsage
+    if (!extendedKeyUsage.contains(EXUpdatesCodeSigningConfiguration.EXUpdatesCodeSigningCertificateExtendedUsageCodeSigningOID)) {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateMissingCodeSigningError
+    }
+    
+    keyId = metadata[EXUpdatesCodeSigningMetadataFields.KeyIdFieldKey] ?? EXUpdatesSignatureHeaderInfo.DefaultKeyId
+    algorithm = try parseCodeSigningAlgorithm(str: metadata[EXUpdatesCodeSigningMetadataFields.AlgorithmFieldKey])
+  }
+  
+  @objc
+  public func createAcceptSignatureHeader() -> String {
+    return "sig, keyid=\"\(keyId)\", alg=\"\(algorithm.rawValue)\""
```

Comment:
escape the keyId and algorithm to prevent any injections

Reply:
Will do in a follow-up PR on both iOS and Android (so that I don't forget to do it on both): https://linear.app/expo/issue/ENG-2834/escape-the-keyid-and-algorithm-to-prevent-any-injections